src/entity/Player.ts

```diff
@@ -122,10 +125,7 @@ export function changeMageType(type: MageType, player?: IPlayer, underworld?: Un
         {
           const upgrade = Upgrade.getUpgradeByTitle(arrowCardId);
           if (upgrade) {
-            underworld.chooseUpgrade(player, upgrade);
-            if (player == globalThis.player) {
-              globalThis.freeSpells.push(arrowCardId);
-            }
```

Comment:
@SoulMuncher I almost missed this.  When this was changed from globalThis.freeSpells to player.freeSpells it still would've been a problem because _every_ client need to have the player.freeSpells.push happen so that all player objects are in parity.  Now getFreeUpgrade handles this.  The player object that is passed in gets the spell and all clients keep their player objects in parity.